File: packages/gbnf-experiment/data/20260909T150124Z_a7cd2ee9/ported_implementation/src/rules-builder/rules-builder-types.ts
import { validateNonEmpty } from '../utils/validate-non-empty';

export const InternalRuleType = {
  CHAR: 'char',
  CHAR_RNG_UPPER: 'char_rng_upper',
  CHAR_ALT: 'char_alt',
  CHAR_NOT: 'char_not',
  ALT: 'alt',
  END: 'end',
  REF: 'ref',
} as const;
export type InternalRuleType =
  (typeof InternalRuleType)[keyof typeof InternalRuleType];

export interface InternalRuleDefChar {
  type: typeof InternalRuleType.CHAR;
  value: number[];
}

export interface InternalRuleDefCharNot {
  type: typeof InternalRuleType.CHAR_NOT;
  value: number[];
}

export interface InternalRuleDefCharAlt {
  type: typeof InternalRuleType.CHAR_ALT;
  value: number;
}

export interface InternalRuleDefCharRngUpper {
  type: typeof InternalRuleType.CHAR_RNG_UPPER;
  value: number;
}

export interface InternalRuleDefReference {
  type: typeof InternalRuleType.REF;
  value: number;
}

export interface InternalRuleDefAlt {
  type: typeof InternalRuleType.ALT;
}

export interface InternalRuleDefEnd {
  type: typeof InternalRuleType.END;
}

export type InternalRuleDef =
  | InternalRuleDefChar
  | InternalRuleDefCharNot
  | InternalRuleDefCharAlt
  | InternalRuleDefCharRngUpper
  | InternalRuleDefReference
  | InternalRuleDefAlt
  | InternalRuleDefEnd;

export type InternalRuleDefCharOrAltChar =
  | InternalRuleDefChar
  | InternalRuleDefCharAlt;

export const internalRuleDefChar = (value: number[]): InternalRuleDefChar => ({
  type: InternalRuleType.CHAR,
  value: [...validateNonEmpty(value)],
});

export const internalRuleDefCharNot = (
  value: number[]
): InternalRuleDefCharNot => ({
  type: InternalRuleType.CHAR_NOT,
  value: [...validateNonEmpty(value)],
});

export const internalRuleDefCharAlt = (
  value: number
): InternalRuleDefCharAlt => ({ type: InternalRuleType.CHAR_ALT, value });

export const internalRuleDefCharRngUpper = (
  value: number
): InternalRuleDefCharRngUpper => ({
  type: InternalRuleType.CHAR_RNG_UPPER,
  value,
});

export const internalRuleDefReference = (
  value: number
): InternalRuleDefReference => ({ type: InternalRuleType.REF, value });

export const internalRuleDefAlt = (): InternalRuleDefAlt => ({
  type: InternalRuleType.ALT,
});

export const internalRuleDefEnd = (): InternalRuleDefEnd => ({
  type: InternalRuleType.END,
});

export const isRuleDefAlt = (rule?: InternalRuleDef): rule is InternalRuleDefAlt =>
  rule?.type === InternalRuleType.ALT;

export const isRuleDefRef = (
  rule?: InternalRuleDef
): rule is InternalRuleDefReference => rule?.type === InternalRuleType.REF;

export const isRuleDefEnd = (rule?: InternalRuleDef): rule is InternalRuleDefEnd =>
  rule?.type === InternalRuleType.END;

export const isRuleDefChar = (rule?: InternalRuleDef): rule is InternalRuleDefChar =>
  rule?.type === InternalRuleType.CHAR;

export const isRuleDefCharNot = (
  rule?: InternalRuleDef
): rule is InternalRuleDefCharNot => rule?.type === InternalRuleType.CHAR_NOT;

export const isRuleDefCharAlt = (
  rule?: InternalRuleDef
): rule is InternalRuleDefCharAlt => rule?.type === InternalRuleType.CHAR_ALT;

export const isRuleDefCharRngUpper = (
  rule?: InternalRuleDef
): rule is InternalRuleDefCharRngUpper =>
  rule?.type === InternalRuleType.CHAR_RNG_UPPER;

File: proposed-projects/round-trip-experiment/reverse/20260909T150124Z_a7cd2ee9/20260910T084451Z_b3565ede/ported_implementation/gbnf/rules_builder/rules_builder_types.py
from ..utils.validate_non_empty import validate_non_empty


class InternalRuleType:
    CHAR = "char"
    CHAR_RNG_UPPER = "char_rng_upper"
    CHAR_ALT = "char_alt"
    CHAR_NOT = "char_not"
    ALT = "alt"
    END = "end"
    REF = "ref"


class InternalRuleDef:
    type: str

    def __repr__(self) -> str:
        value = getattr(self, "value", None)
        if value is None:
            return f"{type(self).__name__}(type={self.type!r})"
        return f"{type(self).__name__}(type={self.type!r}, value={value!r})"


class InternalRuleDefChar(InternalRuleDef):
    type = InternalRuleType.CHAR

    def __init__(self, value: list[int]):
        self.value = list(validate_non_empty(value))


class InternalRuleDefCharNot(InternalRuleDef):
    type = InternalRuleType.CHAR_NOT

    def __init__(self, value: list[int]):
        self.value = list(validate_non_empty(value))


class InternalRuleDefCharAlt(InternalRuleDef):
    type = InternalRuleType.CHAR_ALT

    def __init__(self, value: int):
        self.value = value


class InternalRuleDefCharRngUpper(InternalRuleDef):
    type = InternalRuleType.CHAR_RNG_UPPER

    def __init__(self, value: int):
        self.value = value


class InternalRuleDefReference(InternalRuleDef):
    type = InternalRuleType.REF

    def __init__(self, value: int):
        self.value = value


class InternalRuleDefAlt(InternalRuleDef):
    type = InternalRuleType.ALT


class InternalRuleDefEnd(InternalRuleDef):
    type = InternalRuleType.END


def internal_rule_def_char(value: list[int]) -> InternalRuleDefChar:
    return InternalRuleDefChar(value)


def internal_rule_def_char_not(value: list[int]) -> InternalRuleDefCharNot:
    return InternalRuleDefCharNot(value)


def internal_rule_def_char_alt(value: int) -> InternalRuleDefCharAlt:
    return InternalRuleDefCharAlt(value)


def internal_rule_def_char_rng_upper(value: int) -> InternalRuleDefCharRngUpper:
    return InternalRuleDefCharRngUpper(value)


def internal_rule_def_reference(value: int) -> InternalRuleDefReference:
    return InternalRuleDefReference(value)


def internal_rule_def_alt() -> InternalRuleDefAlt:
    return InternalRuleDefAlt()


def internal_rule_def_end() -> InternalRuleDefEnd:
    return InternalRuleDefEnd()


def is_rule_def_alt(rule: InternalRuleDef | None) -> bool:
    return isinstance(rule, InternalRuleDefAlt)


def is_rule_def_ref(rule: InternalRuleDef | None) -> bool:
    return isinstance(rule, InternalRuleDefReference)


def is_rule_def_end(rule: InternalRuleDef | None) -> bool:
    return isinstance(rule, InternalRuleDefEnd)


def is_rule_def_char(rule: InternalRuleDef | None) -> bool:
    return isinstance(rule, InternalRuleDefChar)


def is_rule_def_char_not(rule: InternalRuleDef | None) -> bool:
    return isinstance(rule, InternalRuleDefCharNot)


def is_rule_def_char_alt(rule: InternalRuleDef | None) -> bool:
    return isinstance(rule, InternalRuleDefCharAlt)


def is_rule_def_char_rng_upper(rule: InternalRuleDef | None) -> bool:
    return isinstance(rule, InternalRuleDefCharRngUpper)
